011-kafka-canary.md

```diff
@@ -0,0 +1,89 @@
+# Strimzi Canary
+
+Implement a Kafka canary tool which will act as an indicator of whether Kafka clusters are operating correctly. This would be achieved by creating a canary topic and periodically producing and consuming events on the topic. 
+
+## Current situation
+
+Currently Strimzi does not have any canary style feature. It does have health checks for the Kafka cluster which can be replaced/augmented with the canary.
+
+## Motivation
+
+The canary would provide the following in indication that the Kafka cluster is operating as expected from a users perspective. I.e. messages can successfully be produced and consumed. Some would be exported to reflect the canary's activity. Initial metrics would included:
+  * records-produced-total
+  * records-consumed-total
+  * produce-error
+  * consume-error
+  * latency
+
+In future the canary can be used by the Kafka Roller to improve health checks. However this is beyond the scope of the current proposed work.
+Proposal
+The canary would be run as a separate pod alongside Strimzi operator. Once the canary topic is created messages will be produced and consumed from it. Messages would be consumed at a rate not faster than 1 per second and possibly as slow as 10 per second. This rate may be configurable. 
+
+The canary will be built using Golang and metrics will be exposed in Prometheus format though a REST API. Currently Sarama is being considered as the client library.
+
+The current plan is to deploy the canary independently  alongside the Strimzi operator. However, in future, consideration should be given to integrating the canary with Strimzi in a manner similar to the Kafka Bridge. I.e. the canary would be specified (optionally) in the Kafka custom resource and then deployed by the Strimzi operator.
+
+### Topic Configuration
+The topic will be configured to have a partition on each broker node and a  well defined replication factor of 3 (which is the most commonly used value).The minimum in-sync replicas should be one less than the number of replicas.
```

Comment:
Some clarifications to make this work even with small clusters.

```suggestion
The topic will be configured to have a partition on each broker node and a replication factor which will be the minimum of _number of Kafka broker nodes_ and _3 (which is the most commonly used value)_. The minimum in-sync replicas will be either 1 in case replication factor is 1 (there is only one broker in the cluster) or one less than the number of replicas.
```